### PR TITLE
fix(gatsby-remark-katex): fix compatibility with remark 13

### LINKED ISSUE #30865: upgrading to gatsby-remark-katex v5.0.0 and/or gatsby-transformer-remark v4.0.0 breaks katex math equation
## upgrading to gatsby-remark-katex v5.0.0 and/or gatsby-transformer-remark v4.0.0 breaks katex math equation

When I use gatsby-remark-katex v4.2.0 and gatsby-transformer-remark v3.2.0, katex math equations are working correctly

aaaa

But when I update to the latest gatsby-remark-katex v5.0.0 and gatsby-transformer-remark v4.0.0, katex math equations no longer work

bbb

### Steps to reproduce

I've created a minimal repo at https://github.com/kimbaudi/gatsby-remark-katex-issue. If you `yarn install && yarn start`, it will work. But if you update to the latest gatsby-remark-katex v5.0.0 and gatsby-transformer-remark v4.0.0, it will no longer work.

### Expected result

What should happen?

updating to the latest gatsby-remark-katex v5.0.0 and gatsby-transformer-remark v4.0.0 should not break katex math equation

### Actual result

What happened.

updating to the latest gatsby-remark-katex v5.0.0 and gatsby-transformer-remark v4.0.0 breaks katex math equation

### Environment

  System:
    OS: Windows 10 10.0.19041
    CPU: (8) x64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz 
  Binaries:
    Node: 14.15.0 - C:\Program Files\nodejs\node.EXE      
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD        
    npm: 7.9.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 89.0.4389.128
    Edge: Spartan (44.19041.906.0), Chromium (89.0.774.76)
  npmPackages:
    gatsby: ^3.3.0 => 3.3.0
    gatsby-remark-katex: ^4.2.0 => 5.0.0
    gatsby-source-filesystem: ^3.3.0 => 3.3.0
    gatsby-transformer-remark: ^3.2.0 => 4.0.0
  npmGlobalPackages:
    gatsby-cli: 3.2.0

### LINKED ISSUE #31427: remark-remark-katex replaces LaTeX math with wrong html
## Description

I am trying to make Gatsby render LaTeX math formulas in my blog posts.

I'm using the standard `gatsby-transformer-remark` with `gatsby-remark-katex` plugin configured as follows:

From `gatsby-config.js` ([full code available here](https://github.com/BurkovBA/BurkovBA.github.io/blob/master/gatsby-config.js)):
```javascript
module.exports = {
  siteMetadata: {
    title: `Personal blog of Boris Burkov`,
    author: `Boris Burkov`,
    description: `A personal blog of Boris Burkov, version 4.1, this time written in Gatsby.`,
    siteUrl: `https://borisburkov.net/`,
    social: {
      telegram: `BorisBurkov`,
    },
  },
  plugins: [
      ...
      {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [
          ...
          {
            resolve: `gatsby-remark-katex`,
            options: {
              // Add any KaTeX options from https://github.com/KaTeX/KaTeX/blob/master/docs/options.md here
              strict: `ignore`
            }
          }
          ...
        ]
      ...
    ],
};
```

In the markdown of my blog posts I'm using both inline math and display-mode math like this:

```md
If we assume for a second that HDD failures were independent, then probability of
a failure of X drives at the same time corresponds to Poisson distribution:

$$
p(failure of 1 HDD) = 10^{-5}
$$

$N(HDD) = 2400$
```

In the resulting html page these math formulas are strangely converted to the following incorrect html:

```html
<div class="math math-display katex">p(failure of 1 HDD) = 10^{-5}</div>
```

Now, if I go to the source code of gatsby-remark-katex/index.js and add check, if it sees my inline math and if it replaces it with proper html, I see that it actually does. I add `    console.log('node.value = ${node.value}')` statement to the code:

`gatsby-remark-katex/index.js`
```javascript
"use strict";

const visit = require(`unist-util-visit`);

const katex = require(`katex`);

const remarkMath = require(`remark-math`);

module.exports = ({
  markdownAST
}, pluginOptions = {}) => {
  visit(markdownAST, `inlineMath`, node => {
    node.type = `html`;
    node.value = katex.renderToString(node.value, {
      displayMode: false,
      ...pluginOptions
    });
    console.log(`node.value = ${node.value}`)
  });
  visit(markdownAST, `math`, node => {
    node.type = `html`;
    node.value = katex.renderToString(node.value, {
      displayMode: true,
      ...pluginOptions
    });
  });
};

module.exports.setParserPlugins = () => [remarkMath];
```

When I say `gatsby develop`, in the console I see the expected html for math formulas, rendered by katex:

```html
node.value = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mo stretchy="false">(</mo><mi>H</mi><mi>D</mi><mi>D</mi><mo
stretchy="false">)</mo><mo>=</mo><mn>2400</mn></mrow><annotation encoding="application/x-tex">N(HDD) = 2400</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut"
style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.10903em;">N</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span
class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="mclose">)</span><span class="mspace"
style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut"
style="height:0.64444em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord">4</span><span class="mord">0</span><span class="mord">0</span></span></span></span>
```

Problem is that html wouldn't show up in the resulting page. It wouldn't appear in GraphQL of the page in `html` or `htmlAST` field either, if I go to query the GraphiQL for it:

Screenshot 2021-05-14 at 19 37 23 

Here's the page output, note the un-rendered formulae:

Screenshot 2021-05-14 at 19 55 00

### Steps to reproduce

Full source code of my Gatsby 3-based blog is available here: https://github.com/BurkovBA/BurkovBA.github.io.git.

```
git clone https://github.com/BurkovBA/BurkovBA.github.io.git
nvm use 15
npm install
npm run develop
browser http://localhost:8000/2021-03-11-1/
```

### Expected result

I expected LaTeX formulas rendered in the lower part of my latest blog post: http://localhost:8000/2021-03-11-1/ as the following html:

```html
<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mo stretchy="false">(</mo><mi>H</mi><mi>D</mi><mi>D</mi><mo
stretchy="false">)</mo><mo>=</mo><mn>2400</mn></mrow><annotation encoding="application/x-tex">N(HDD) = 2400</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut"
style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.10903em;">N</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span
class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="mclose">)</span><span class="mspace"
style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut"
style="height:0.64444em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord">4</span><span class="mord">0</span><span class="mord">0</span></span></span></span>
```

### Actual result

LaTeX is not rendered, but just replaced with pretty useless divs:

```html
<div class="math math-display">p(failure of 1 HDD) = 10^{-5}</div>
```

### Environment

```
  System:
    OS: macOS 11.1
    CPU: (8) arm64 Apple M1
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 15.14.0 - ~/.nvm/versions/node/v15.14.0/bin/node
    npm: 7.12.1 - ~/.nvm/versions/node/v15.14.0/bin/npm
  Languages:
    Python: 3.9.2 - /Users/burkov/miniforge3/bin/python
  Browsers:
    Safari: 14.0.2
  npmPackages:
    gatsby: ^3.4.2 => 3.4.2 
    gatsby-plugin-feed: ^3.4.0 => 3.4.0 
    gatsby-plugin-google-analytics: ^3.4.0 => 3.4.0 
    gatsby-plugin-image: ^1.4.1 => 1.4.1 
    gatsby-plugin-manifest: ^3.4.0 => 3.4.0 
    gatsby-plugin-offline: ^4.4.0 => 4.4.0 
    gatsby-plugin-react-helmet: ^4.4.0 => 4.4.0 
    gatsby-plugin-sass: ^4.4.0 => 4.4.0 
    gatsby-plugin-sharp: ^3.4.2 => 3.4.2 
    gatsby-plugin-styled-components: ^4.4.0 => 4.4.0 
    gatsby-plugin-typography: ^3.4.0 => 3.4.0 
    gatsby-remark-copy-linked-files: ^4.1.0 => 4.1.0 
    gatsby-remark-embed-gist: ^1.2.1 => 1.2.1 
    gatsby-remark-images: ^5.1.0 => 5.1.0 
    gatsby-remark-katex: ^5.1.0 => 5.1.0 
    gatsby-remark-prismjs: ^5.1.0 => 5.1.0 
    gatsby-remark-responsive-iframe: ^4.1.0 => 4.1.0 
    gatsby-remark-smartypants: ^4.1.0 => 4.1.0 
    gatsby-source-filesystem: ^3.4.0 => 3.4.0 
    gatsby-transformer-remark: ^4.1.0 => 4.1.0 
    gatsby-transformer-sharp: ^3.4.0 => 3.4.0 
  npmGlobalPackages:
    gatsby-cli: 3.4.1
```

No `flags` inside `gatsby-config.js`.

### PR DESCRIPTION
**Pull request by BurkovBA**
## Description

gatsby-remark-katex plugin as of v3.5 was not rendering math (both inline and display mode).

### Documentation

Discussed in detali in #30865

## Related Issues

Fixes #30865
Fixes #31427

### CODE REVIEW

**Comment by BurkovBA** [comment]
No, I don't think it's a breaking change, the plugin interface stays the
same.

The mechanism of plugin inner functioning changes significantly, though.

ср, 26 мая 2021 г., 18:56 Vladimir Razuvaev ***@***.***>:

**Comment by BurkovBA** [comment]
Dear maintainers, do I have to do anything to get things going with this PR?

How to trigger Circle CI here? Can anyone review this PR?

File: packages/gatsby-remark-katex/src/index.js [hunk 1]
```diff
@@ -1,22 +1,26 @@
 const visit = require(`unist-util-visit`)
```
Comment (kimbaudi) [inline packages/gatsby-remark-katex/src/index.js:1]: would it be okay to remove usage of `const parseHtml` and `const renderedKatexString` and just keep the previous form? I understand you want to remove duplicate code, but the previous form is easier to understand.

```js
const visit = require(`unist-util-visit`)
const unified = require(`unified`)
const parse = require(`rehype-parse`)
const katex = require(`katex`)
const remarkMath = require(`remark-math`)

module.exports = ({ markdownAST }, pluginOptions = {}) => {
  visit(markdownAST, `inlineMath`, (node) => {
    node.data.hChildren = unified()
      .use(parse, { fragment: true, position: false })
      .parse(
        katex.renderToString(node.value, {
          displayMode: false,
          ...pluginOptions,
        })
      ).children
  })

  visit(markdownAST, `math`, (node) => {
    node.data.hChildren = unified()
      .use(parse, { fragment: true, position: false })
      .parse(
        katex.renderToString(node.value, {
          displayMode: true,
          ...pluginOptions,
        })
      ).children
  })
}

module.exports.setParserPlugins = () => [remarkMath]

```

**Review by kimbaudi — APPROVED** [review]
i'm not authorized to merge this PR, but I reviewed your code and it looks good to me. i left a suggestion, but you're free to ignore it. Thanks for the PR!

**Comment by kimbaudi** [comment]
> Technically, it is a breaking change, right?

The breaking change is from [upgrading to remark 13](https://github.com/gatsbyjs/gatsby/pull/29678/) where `gatsby-remark-katex` plugin updated `remark-math` dependency from version `1.0.6` to version `4.0.0` (which requires remark 13). See https://github.com/gatsbyjs/gatsby/pull/29678/files#diff-1d1d8914cc8969d5a300772e4f734925b1e868c417f50bb0865be9488e3b84f9

This PR is a fix to that breaking change. If this is in regards to semver for this pacakge, I think it should be fine as a patch version update (`5.4.1-next.0`) instead of a major version update (`5.5.0-next.0`)

**Comment by kimbaudi** [comment]
I still think this PR needs a decent unit test since the unit test that is in place didn't help detect this bug prior to release.

**Comment by BurkovBA** [comment]
@kimbaudi Updated the code according to your suggestions. Bumped the version to 5.4.1-next.1. Updated a scoped package @burkov/gatsby-remark-katex, tested against my blog, works fine.

**Review by kimbaudi — DISMISSED** [review]

File: packages/gatsby-remark-katex/CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,10 @@
 All notable changes to this project will be documented in this file.
 See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
 
+# [5.4.1-next.1](https://github.com/gatsbyjs/gatsby/compare/gatsby-remark-katex@5.4.0-next.0...gatsby-remark-katex@5.4.1-next.1) (2021-05-29)
+
+**Note:** Version bump only for package gatsby-remark-katex
```
Comment (kimbaudi) [inline packages/gatsby-remark-katex/CHANGELOG.md:8]: maybe change this to something like:

```raw
### Features

- fix issue with katex not rendering after upgrade to remark 13 ([#30865](https://github.com/gatsbyjs/gatsby/issues/30865))
```

File: packages/gatsby-remark-katex/package.json [hunk 3]
```diff
@@ -29,7 +32,7 @@
   "main": "index.js",
   "peerDependencies": {
     "gatsby": "^3.0.0-next.0",
-    "katex": "^0.12.0"
+    "katex": "^0.13.3"
```
Comment (kimbaudi) [inline packages/gatsby-remark-katex/package.json:35]: current version is `^0.13.11`, but I'm not sure if it is convention to keep it at the lowest patch version (i.e. `^0.13.0`). Maybe change it to  `^0.13.11`? Or just leave it as is.

**Review by kimbaudi — COMMENTED** [review]
i think the changelog should make a note about this PR being a bug fix. Other than that, looks good to me.

File: packages/gatsby-remark-katex/CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,10 @@
 All notable changes to this project will be documented in this file.
 See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
 
+# [5.4.1-next.1](https://github.com/gatsbyjs/gatsby/compare/gatsby-remark-katex@5.4.0-next.0...gatsby-remark-katex@5.4.1-next.1) (2021-05-29)
```
Comment (wardpeet) [inline packages/gatsby-remark-katex/CHANGELOG.md:6]: Please remove this file :)

File: packages/gatsby-remark-katex/src/__tests__/__snapshots__/index.js.snap.old [hunk 5]
```diff
@@ -0,0 +1,189 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`remark katex plugin renders double $ inlineMath node properly 1`] = `
```
Comment (wardpeet) [inline packages/gatsby-remark-katex/src/__tests__/__snapshots__/index.js.snap.old:3]: Shouldn't we remove this?

File: packages/gatsby-remark-katex/CHANGELOG.md (hunk 4 above)
Comment (BurkovBA) [reply packages/gatsby-remark-katex/CHANGELOG.md]: Removed the changelog line.

File: packages/gatsby-remark-katex/src/__tests__/__snapshots__/index.js.snap.old (hunk 5 above)
Comment (BurkovBA) [reply packages/gatsby-remark-katex/src/__tests__/__snapshots__/index.js.snap.old]: Removed the old jest snapshot.

**Review by vladar — APPROVED** [review]
LGTM. I've tried it locally and it worked for me. @BurkovBA thanks so much for the fix! 💜

**Comment by ko1imai** [comment]
[gatsby-remark-katex] [gatsby-plugin-mdx]
I'm wondering if I should create a new issue or use this existing issue (#31596). I imagine this fix would work under gatsby-tranformer-remark which uses remark 13. Instead, however, I faced an issue with Gatsby-plugin-mdx which depends on remark 10. Actually, I got a warning message during gatsby develop saying that "warn [remark-math] Warning: please upgrade to remark 13 to use this plugin". I would appreciate it if you could consider updating remark 13 used in Gatsby-plugin-mdx. Or, alternatively please consider to make this fix works also with remark 10, an older version. This might be a side effect of this issue, I got another error message, saying as follows:
```
warn Error persisting state: function math() {
  var data = this.data()
  /* istanbul ignore next - old remark. */
  if (
    !warningIssue...<omitted>...
} could not be cloned.
```

Best regards, Koichi Imai (in Japan)

**Comment by nikoladev** [comment]
@ko1imai Generally it's best to open new issues for these types of things.

But [here's the official answer to your question][link]. Basically, only after version 2 of https://github.com/mdx-js/mdx is released can `gatsby-plugin-mdx` upgrade to `remark@13`.

But to be fair, I don't know when version 2 of mdx will be released. There isn't too much progress on that in the mdx repository 

[link]: https://github.com/gatsbyjs/gatsby/discussions/30385#discussioncomment-575787

**Pull request merged**